### PR TITLE
server: Introduce ServerFactoryContext

### LINKED ISSUE #8303: rds: make RouteConfigProvider shared among ListenerImpls, or move factory_context_ outside of Listener.
*Title*: To resolve issues like #7923, or migrate RDS to config-provider-framework, we need to make sure RdsRouteConfigProvider "Listener independent".

*Description*:
> At present a RdsRouteConfigProviderImpl is bound to a listenerImpl, as it holds a FactoryContext references to the ListenerImpl. 

If we want to save memory on dedup the copy of RDS route config providers (e.g., #7923 ), we would need to make sure the referenced FactoryContext outlives a ListenerImpl.

Besides that, a RDS Subscription is held globally (weak_ptr) by the SingletonManager[RdsConfigProviderManager],which could be shared between Listeners, but the RDSSubscription also holds a reference to the factoryContext (in this case, ListenerImpl). If LDS updates a listener A, a live RDSSubscription owned by another listener B may crash Envoy when it tries to read from the destructed FactoryContext(ListenerImpl). VHDS implementation has a TODO that stated this concern:
https://github.com/envoyproxy/envoy/blob/master/source/common/router/rds_impl.cc#L113

Proposal: 

One way to make the factoryContext referenced by router/... is to pass a Server::CommonFactoryContext (e.g. the Server::Instance ) instead of a Server::FactoryContext (only implemented by ListenerImpl) into the router/...

Right now, most of the APIs desired by objects under router/... are actually from CommonFactoryContext, except for several places that are kinda Listener specific: 

- initManager(), which is used to initiate a Subscription, but as SRDS(also VHDS, see issue#7617) comes into Envoy, we need something like the noop_init_manager in scoped_rds.cc 
We could possibly pass the current ListernerImpl::initManager() along with the CommonFactoryContext(a.k.a the Server::Instance) to router/... As it's only required at Subscription Start time.
- RouteEntryImplBase::per_filter_configs_ 
  This is the place where users could insert their own per-filter-per-route config data, a FactoryContext is passed in to translate the opaque protobuf::Struct into C++  RouteSpecificFilterConfig object. In the upstream codebase, this parameter is not used in any translation so far. But there is a risk some users' translation may fail if they depends on some per-listener data during the translation, if we decide to move forward by passing a CommonFactoryContext into router/... instead of the ListenerImpl itself.

### PR DESCRIPTION
**Pull request by lambdai**
Description:
Introduce ServerFactoryContext which is an implementation of CommonFactoryContext. Comparing to ListenerFactoryContext, ServerFactoryContext has a wider lifetime.
This is the prerequisite to share rds configuration among listeners.

Also experimenting move messageValidateVisitor out of CommonFactoryContext. The benefit is that now Server::Instance perfectly implements CommonFactoryContext. 
* ServerFactoryContext doesn't implement the function. The test shows it is never invoked.
* Will put `messageValidateVisitor` into `FactoryContext`: this piece of work will be done in follow up PR

Notes to reviewers:
~~clang-tidy error is not introduced in this PR.  Create https://github.com/envoyproxy/envoy/issues/8509 to track~~
~~Will fix once this PR is merged~~

This PR is extracted from https://github.com/envoyproxy/envoy/pull/8454

Risk Level: 
Testing: 
Docs Changes:
Release Notes:
Fixes #8303

### CODE REVIEW

**Comment by lambdai** [comment]
Not fixing all the tests yet. Will resume tomorrow

**Comment by lambdai** [comment]
clang_tidy passed!

**Comment by lambdai** [comment]
Also CC @mattklein123 as in the #8303

**Comment by lambdai** [comment]
gentle ping @mattklein123 @htuch

File: include/envoy/server/filter_config.h [hunk 1]
```diff
@@ -111,6 +113,7 @@ class FactoryContext : public virtual CommonFactoryContext {
 public:
   ~FactoryContext() override = default;
 
+  virtual ServerFactoryContext& getServerFactoryContext(bool is_dynamic) const PURE;
```
Comment (mattklein123) [inline include/envoy/server/filter_config.h]: method/param comments please, also nit newline after definition.

File: include/envoy/server/filter_config.h [hunk 2]
```diff
@@ -102,6 +102,8 @@ class CommonFactoryContext {
   virtual Api::Api& api() PURE;
 };
 
+class ServerFactoryContext;
```
Comment (mattklein123) [inline include/envoy/server/filter_config.h]: Is this needed due to circular include?

File: include/envoy/server/instance.h [hunk 3]
```diff
@@ -224,6 +224,12 @@ class Instance {
    *         messages.
    */
   virtual ProtobufMessage::ValidationContext& messageValidationContext() PURE;
+
+  /**
+   * @param is_dynamic true if the validation visitor is for dynamic config
```
Comment (mattklein123) [inline include/envoy/server/instance.h]: Can we name the parameter something more clear for what this does? It's not super clear to me what is different about the returned context so perhaps more comments?

File: source/common/local_info/local_info_impl.h [hunk 4]
```diff
@@ -10,8 +10,9 @@ namespace LocalInfo {
 class LocalInfoImpl : public LocalInfo {
 public:
   LocalInfoImpl(const envoy::api::v2::core::Node& node,
-                Network::Address::InstanceConstSharedPtr address, const std::string zone_name,
-                const std::string cluster_name, const std::string node_name)
+                const Network::Address::InstanceConstSharedPtr& address,
+                const std::string zone_name, const std::string cluster_name,
```
Comment (mattklein123) [inline source/common/local_info/local_info_impl.h]: While you are in here, can you either pass these by const ref or make them non-const and move them?

File: source/common/router/rds_impl.cc [hunk 5]
```diff
@@ -174,7 +178,9 @@ RdsRouteConfigProviderImpl::RdsRouteConfigProviderImpl(
     RdsRouteConfigSubscriptionSharedPtr&& subscription,
     Server::Configuration::FactoryContext& factory_context)
     : subscription_(std::move(subscription)),
-      config_update_info_(subscription_->routeConfigUpdate()), factory_context_(factory_context),
+      config_update_info_(subscription_->routeConfigUpdate()),
+      factory_context_(factory_context.getServerFactoryContext(/*is_dynamic=*/true)),
+      // provider_init_manager_(factory_context.initManager()),
```
Comment (mattklein123) [inline source/common/router/rds_impl.cc]: ?

File: source/common/router/scoped_rds.cc [hunk 6]
```diff
@@ -267,7 +267,8 @@ void ScopedRdsConfigSubscription::onRdsConfigUpdate(const std::string& scope_nam
   auto new_scoped_route_info = std::make_shared<ScopedRouteInfo>(
       envoy::api::v2::ScopedRouteConfiguration(iter->second->configProto()),
       std::make_shared<ConfigImpl>(rds_subscription.routeConfigUpdate()->routeConfiguration(),
-                                   factory_context_, false));
+                                   factory_context_.getServerFactoryContext(/*is_dynamic=*/false),
```
Comment (mattklein123) [inline source/common/router/scoped_rds.cc]: Shouldn't this be dynamic? TBH I'm not a fan of passing this type of param everywhere, but let's discuss in the interface so I can learn more what this is for.

File: include/envoy/server/filter_config.h [hunk 7]
```diff
@@ -186,6 +189,12 @@ class FactoryContext : public virtual CommonFactoryContext {
   virtual OptProcessContextRef processContext() PURE;
 };
 
+// ServerFactoryContext should live at least as long as lifetime of server.
```
Comment (htuch) [inline include/envoy/server/filter_config.h]: Maybe add a little more explanation of what it is, beyond just lifetime.

File: source/server/server.cc [hunk 8]
```diff
@@ -74,7 +74,9 @@ InstanceImpl::InstanceImpl(
       mutex_tracer_(options.mutexTracingEnabled() ? &Envoy::MutexTracerImpl::getOrCreateTracer()
                                                   : nullptr),
       grpc_context_(store.symbolTable()), http_context_(store.symbolTable()),
-      process_context_(std::move(process_context)), main_thread_id_(std::this_thread::get_id()) {
+      process_context_(std::move(process_context)), main_thread_id_(std::this_thread::get_id()),
+      dynamic_server_context_(*this, /*is_dynamic=*/true),
```
Comment (htuch) [inline source/server/server.cc]: I don't think we need two completely different contexts just for message validation.

File: source/server/server.h [hunk 9]
```diff
@@ -141,6 +141,35 @@ class RunHelper : Logger::Loggable<Logger::Id::main> {
   Event::SignalEventPtr sig_hup_;
 };
 
+class ServerFactoryContextImpl : public Configuration::ServerFactoryContext {
```
Comment (htuch) [inline source/server/server.h]: So, the way I envisaged this playing out is that the `ListenerImpl` factory context should now be massively slimmed down. It doesn't need to be offering most of the methods it has, they can come via the server factory context. But, I don't see a lot of change to listener manager in this PR.

Comment (lambdai) [reply source/server/server.h]: Sorry I don't fully get your idea.
Are you saying ListenerFactoryContext should exclude what ServerFactoryContext provides? Or are you suggesting ListenerImpl should offer the `ServerFactoryContext` functions through a impl of `ServerFactoryContext`?

**Comment by lambdai** [comment]
Chatted with @htuch 
The goal is not to introduce is_dynamic arg and the messageValidationVisitor in ServerFactoryContext. So that ServerFactoryContext is a subset of server::Instance.

 The plan is to 
1. messageValidationVisitor() should not be used in ServerFactoryContext. Mark it NOT_IMPLEMENT. Provide a visitor ref if it is needed.
2. Moving messageValidationVisitor out of CommonFactoryContext in the follow up PR.

**Comment by lambdai** [comment]
It's probably flaky test since I couldn't reproduce
/azp run

File: source/common/router/rds_impl.cc [hunk 10]
```diff
@@ -41,8 +41,12 @@ StaticRouteConfigProviderImpl::StaticRouteConfigProviderImpl(
     const envoy::api::v2::RouteConfiguration& config,
     Server::Configuration::FactoryContext& factory_context,
     RouteConfigProviderManagerImpl& route_config_provider_manager)
-    : config_(new ConfigImpl(config, factory_context, true)), route_config_proto_{config},
-      last_updated_(factory_context.timeSource().systemTime()),
+    :
```
Comment (mattklein123) [inline source/common/router/rds_impl.cc]: nit: a bunch of extra new lines in here.

File: source/server/listener_manager_impl.h [hunk 11]
```diff
@@ -232,6 +232,9 @@ class ListenerImpl : public Network::ListenerConfig,
                ProtobufMessage::ValidationVisitor& validation_visitor);
   ~ListenerImpl() override;
 
+  Configuration::ServerFactoryContext& getServerFactoryContext() const override {
+    return parent_.server_.serverFactoryContext();
+  }
```
Comment (mattklein123) [inline source/server/listener_manager_impl.h]: nit: new line after this line.

File: source/server/server.h [hunk 12]
```diff
@@ -141,6 +141,32 @@ class RunHelper : Logger::Loggable<Logger::Id::main> {
...
+
+  Upstream::ClusterManager& clusterManager() override { return server_.clusterManager(); }
+  Event::Dispatcher& dispatcher() override { return server_.dispatcher(); }
+  const LocalInfo::LocalInfo& localInfo() const override { return server_.localInfo(); }
+  Envoy::Runtime::RandomGenerator& random() override { return server_.random(); }
+  Envoy::Runtime::Loader& runtime() override { return server_.runtime(); }
+  Stats::Scope& scope() override { return *server_scope_; }
+  Singleton::Manager& singletonManager() override { return server_.singletonManager(); }
+  ThreadLocal::Instance& threadLocal() override { return server_.threadLocal(); }
+  // TODO(lambdai): remove messageValidationVisitor from ServerFactoryContext
```
Comment (mattklein123) [inline source/server/server.h]: Can this just be done as part of this change?

**Comment by lambdai** [comment]
@mattklein123 Thanks!
The next PRs will be out pretty soon. With reasonable size and avoid you reviewers going though this large but plain PR.

**Comment by lambdai** [comment]
Oops. I am resolving the conflict
I suspect it's due to another PR by me...

File: source/server/server.h (hunk 12 above)
Comment (lambdai) [reply source/server/server.h]: Please allow me to address it in a follow up PR

**Comment by lambdai** [comment]
@mattklein123 
The last commit is about merging 
the listener_impl.h split PR
and 
forward proxy which uses ConfigPerRoute

**Comment by lambdai** [comment]
@mattklein123 I think the failed envoy-macos is pure test infra failure. Is it good to merge? If not, should i merge master and retrigger the tests? Thanks!

**Comment by mattklein123** [comment]
@lambdai the change LGTM but @htuch had previously reviewed so I would like him to verify and merge. Thank you!

**Comment by lambdai** [comment]
@mattklein123 Thank you for the update!

File: source/server/server.h (hunk 12 above)
Comment (htuch) [reply source/server/server.h]: OK, but this really should be removed in the next PR in this space, as it's a problematic pattern, we now have multiple validation visitors being effectively delivered to the per-route filter factory method.

**Pull request merged**